### PR TITLE
exo firewall

### PR DESCRIPTION
**Pull request by pierre-emmanuelJ**
Signed-off-by: Pierre-Emmanuel Jacquier <pierre-emmanuel.jacquier@epitech.eu>

### CODE REVIEW

File: cmd/exo/cmd/firewall_add.go [hunk 1]
```diff
@@ -42,49 +65,183 @@ func firewallAddRun(cmd *cobra.Command, args []string) {
...
+	}
+	return sgRes, nil
+}
+
+func getDefaultRule(ruleName string) (*egoscale.AuthorizeSecurityGroupIngress, error) {
+	ruleName = strings.ToLower(ruleName)
+	if ruleName == "ping" {
+		return &egoscale.AuthorizeSecurityGroupIngress{
+			Protocol:    "icmp",
+			CidrList:    []string{"0.0.0.0/0"},
```
Comment (greut) [inline cmd/exo/cmd/firewall_add.go]: I'd create some constants for the open IPv4 and IPv6 CIDR values.

File: cmd/exo/cmd/vm_create.go [hunk 2]
```diff
@@ -150,13 +149,29 @@ func getCommaflag(p string) []string {
...
+	for _, arg := range args {
+		if arg == "" {
+			continue
+		}
+		res = append(res, strings.TrimSpace(arg))
 	}
 
-	ps := strings.FieldsFunc(p, f)
+	//XXX c'ant keep this code it will be not working with e.g (ip, cidr, name with special caracters like -, =, ;, /, :)
+	// it's working just for simple case aplhanumeric
```
Comment (greut) [inline cmd/exo/cmd/vm_create.go]: then, remove it.

File: cmd/exo/cmd/firewall_add.go (hunk 1 above)
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/firewall_add.go]: 👍

File: cmd/exo/cmd/firewall.go [hunk 3]
```diff
@@ -75,7 +75,7 @@ func getMyCIDR(isIpv6 bool) (*net.IPNet, error) {
 	dnsServer := ""
 
 	if isIpv6 {
-		dnsServer = "resolver1.ipv6-sandbox.opendns.com"
+		dnsServer = "resolver2.ipv6-sandbox.opendns.com"
```
Comment (greut) [inline cmd/exo/cmd/firewall.go]: uhu

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/firewall.go]: 😂

**Comment by greut** [comment]
I've create an `all` rule.

```
        {
          "cidr": "::/0",
          "protocol": "all",
          "ruleid": "799584ba-e5c4-4561-b3bf-390a6c6e3f03"
        }

```

It should not display any value for the _port_ column.

|  TYPE   |     SOURCE     | PROTOCOL |   PORT   | DESCRIPTION |                  ID                  |
|---------|----------------|----------|----------|-------------|--------------------------------------|
| INGRESS | CIDR 0.0.0.0/0 | tcp      | 22 (ssh) | SSH         | fc03b5b1-1d15-4933-99c3-afa0b8f2ab25 |
|         | CIDR ::/0      | all      | 0        |             | 799584ba-e5c4-4561-b3bf-390a6c6e3f03 |

**Comment by pierre-emmanuelJ** [comment]
so it is bad ?

**Comment by pierre-emmanuelJ** [comment]
it should display `all protocol` ?

**Comment by pierre-emmanuelJ** [comment]
it should display `all protocol` ?

**Comment by greut** [comment]
definitively not zero.

**Comment by pierre-emmanuelJ** [comment]
yep yep

**Comment by greut** [comment]
an issue with `omitempty` ?

**Comment by greut** [comment]
Nothing we can do here... the API says zero.

```
        {
          "cidr": "::/0",
          "endport": 0,
          "protocol": "all",
          "ruleid": "6e3208a8-ed1e-4512-8548-f894718ca9ab",
          "startport": 0,
          "tags": []
        }
```

the `omitempty` is causing us trouble here...

**Comment by greut** [comment]
so everything is fine, do not do anything :-)

File: cmd/exo/cmd/firewall.go [hunk 4]
```diff
@@ -62,6 +69,37 @@ func getSecuGrpWithNameOrID(cs *egoscale.Client, name string) (*egoscale.Securit
...
+		PreferGo: true,
+	}
+
+	ip, err := resolver.LookupIPAddr(context.Background(), "myip.opendns.com")
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ip) < 1 {
+		return nil, fmt.Errorf("Invalide IP adress")
```
Comment (greut) [inline cmd/exo/cmd/firewall.go]: french?

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/firewall.go]: Invalid IP adress

File: cmd/exo/cmd/firewall_add.go [hunk 5]
```diff
@@ -32,103 +37,275 @@ func firewallAddRun(cmd *cobra.Command, args []string) {
...
 func init() {
 	firewallAddCmd.Run = firewallAddRun
 
-	firewallAddCmd.Flags().BoolP("ipv6", "6", false, "Add rule for any IPv6 source")
-	firewallAddCmd.Flags().BoolP("my-ip", "m", false, "Add rule only for my IP as a source")
+	firewallAddCmd.Flags().BoolP("ipv6", "6", false, "Set ipv6 on default rules or on --my-ip")
+	firewallAddCmd.Flags().BoolP("my-ip", "", false, "Set CIDR for my ip")
+	firewallAddCmd.Flags().BoolP("engress", "e", false, "By default rule is INGRESS (set --engress to have ENGRESS rule)")
+	firewallAddCmd.Flags().StringP("protocol", "p", "", "Rule Protocol available [tcp, udp, icmp, icmpv6, ah, esp, gre]")
+	firewallAddCmd.Flags().StringP("cidr", "c", "", "Rule Cidr [CIDR 0.0.0.0/0,192.168.0.0/16...]")
```
Comment (greut) [inline cmd/exo/cmd/firewall_add.go]: `::/0` is probably more useful than 192.168.x.x (a non-routable IP)

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/firewall_add.go]: yes

**Comment by pierre-emmanuelJ** [comment]
all is good for you ?

File: cmd/exo/cmd/firewall.go [hunk 6]
```diff
@@ -62,6 +69,37 @@ func getSecuGrpWithNameOrID(cs *egoscale.Client, name string) (*egoscale.Securit
...
+		PreferGo: true,
+	}
+
+	ip, err := resolver.LookupIPAddr(context.Background(), "myip.opendns.com")
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ip) < 1 {
+		return nil, fmt.Errorf("Invalid IP adress")
```
Comment (greut) [inline cmd/exo/cmd/firewall.go]: _address_

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/firewall.go]: OMG !

**Pull request merged**